src/box/lua/config/applier/credentials.lua

```diff
@@ -277,6 +282,190 @@ local function privileges_add_defaults(name, role_or_user, intermediate)
+    -- instance:
+    --
+    -- Default roles:
+    -- * super
+    -- * public
+    -- * replication
+    --
+    -- Default users:
+    -- * guest
+    -- * admin
+    --
+    -- These roles and users have according privileges pre-granted by design.
+    -- Credentials applier adds such privileges with `priviliges_add_default()`
+    -- when syncing. So, for the excessive (non-default) privs to be removed,
+    -- these roles and users must be present inside configuration at least in a
+    -- form of an empty table. Otherwise, the privileges will be left unchanged,
+    -- similar to all used-defined roles and users.
+
+    credentials.roles = credentials.roles or {}
+    credentials.roles['super'] = credentials.roles['super'] or {}
+    credentials.roles['public'] = credentials.roles['public'] or {}
+    credentials.roles['replication'] = credentials.roles['replication'] or {}
+
+    credentials.users = credentials.users or {}
+    credentials.users['guest'] = credentials.users['guest'] or {}
+    credentials.users['admin'] = credentials.users['admin'] or {}
+
+    return credentials
+end
+
+-- }}} Sync helpers
+
+-- {{{ Triggers for grants after obj creation
+
+-- The table contains flags for objects that should be synced in the trigger:
+-- [obj_type][obj_name] = true/nil
+local on_creation_flags = {
+    ['space'] = {},
+    ['function'] = {},
+    ['sequence'] = {},
+}
+
+-- Info about objects is stored inside system spaces
+-- box.space._space/_func/_sequence. Translate these names
+-- to obj_type values of space/function/sequence.
+local function translate_obj_type(obj_type)
+    if obj_type == '_space' then
+        return 'space'
+    elseif obj_type == '_func' then
+        return 'function'
+    elseif obj_type == '_sequence' then
+        return 'sequence'
+    end
+end
+
+local sync_privileges
+
+-- Iterate through objects and sync when according flag is set.
+local function on_commit_trigger(iterator)
+    log.debug('credentials.apply: on_commit_trigger() started')
+    for _, old_obj, new_obj, space_id in iterator() do
+        if box.space[space_id] == nil then
+            goto skip
+        end
+
+        local obj_type = box.space[space_id].name
+        obj_type = translate_obj_type(obj_type)
+
+        -- Check if the request is for space/function/sequence.
+        if obj_type == nil then
+            goto skip
+        end
+        if new_obj == nil then
+            goto skip
+        end
+
+        -- Check if the request is for obj creation or obj rename.
+        if old_obj == nil or old_obj.name ~= new_obj.name then
+            local obj_name = new_obj.name
+
+            -- Check that the object is present inside config.
+            if on_creation_flags[obj_type][obj_name] then
+                local credentials = get_credentials(config)
+                sync_privileges(credentials, {name = obj_name, type = obj_type})
+            end
```

Comment:
Gleb highlighted that DQL/DML/DDL is forbidden in on_commit triggers (see a49c8c02a2f6b7620db12988ec7ee0369076d4d4). So, it seems, we're obligated to implement it in some other way. (This comment is to track the problem.)

Reply:
Now the sync is performed inside a dedicated fiber. The `on_commit` trigger just schedules a sync for the object via `sync_tasks` fiber channel.